pkg/prometheus/common.go

```diff
@@ -439,3 +443,32 @@ func BuildWebconfig(
 
 	return webConfig.GetMountParameters()
 }
+
+// If a ServiceName is specified, check that the service exists in the namespace. If it does not exist, fail
+// the reconciliation.
+// Also, ensure that the Prometheus instance is selected by the service. If not, fail the reconciliation.
+func CheckCustomService(ctx context.Context, serviceName string, prometheusNS string, prometheusName string, svcClient clientv1.ServiceInterface, selectorLabels map[string]string) error {
```

Comment:
If the function receives a `PrometheusInterface` argument + svcClient + selectorLabels, it could do:
* Verify that the service exists and matches the pods if p.Spec.ServiceName isn't nil
* Otherwise create/update the prometheus-operated service.

Reply:
I agree, that way we could do everything related to the service creation within this one function.